# Handwriting Recognition Using Logistic Regression

## Identify the Problem

Identifying handwritten characters is a common problem in computer vision applications. For example, the USPS uses algorithms to recognize handwritten addresses on envelopes to enable automated mail sorting. Solving this problem using traditional strategies (i.e. edge detection) can be challenging. Can we use data science techniques to recognize handwritten digits (e.g. 0, 1, 2, ... , 9) with a reasonable degree of accuracy?

## Acquire the Data
The MNIST data set is a collection of 70,000 labeled images of handwritten digits. http://yann.lecun.com/exdb/mnist/

We use a pre-processed and pickled version of the dataset made by deeplearning.net. The pickled version can be downloaded from: http://deeplearning.net/data/mnist/mnist.pkl.gz

In [43]:
import os
from sklearn import linear_model, cross_validation, metrics
import pandas as pd
import matplotlib.pyplot as plt
import cPickle
import gzip

In [44]:
# for convinience, we use a pickled version of the MNIST dataset
f = gzip.open('mnist.pkl.gz', 'rb')
train_set, valid_set, test_set = cPickle.load(f)
f.close()

## Parse, Mine, and Refine the Data

'train_set' is a tuple. The first element is an array where each row represents an image. An image consists of 28x28 pixels where each pixel is represented as a float.

The second element is an array of the corresponding labels for each image.

We can extract a single training example and plot it using pyplot.

<img src='assets\example_image.png'>

In [45]:
# PLOT IMAGE
# image is rc/binary format
plt.rc('image', cmap='binary')
# reshape first training image into 28x28 format
plt.matshow((train_set[0][0]).reshape(28, 28))
# label of first training image
plt.title("Label: %d" % train_set[1][0])
plt.show()

The 70,000 images are split as follows:
* 50,000 are put into the training set
* 10,000 are put into the validation set
* 10,000 are put into the test set

In [46]:
print("Length of training set: %d" % len(train_set[0]))
print("Length of validation set: %d" % len(valid_set[0]))
print("Length of test set: %d" % len(test_set[0]))

Length of training set: 50000
Length of validation set: 10000
Length of test set: 10000


## Build a Model

Since this is a classification task, we use train a logistic regression model to classify images into one of 10 categories: 0, 1, 2, ..., 9.

In [47]:
model = linear_model.LogisticRegression()
model.fit(train_set[0], train_set[1])

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [48]:
pred = model.predict(test_set[0])
print metrics.accuracy_score(test_set[1], pred)
print zip(pred[0:100], test_set[1][0:100])

0.9198
[(7, 7), (2, 2), (1, 1), (0, 0), (4, 4), (1, 1), (4, 4), (9, 9), (6, 5), (9, 9), (0, 0), (6, 6), (9, 9), (0, 0), (1, 1), (5, 5), (9, 9), (7, 7), (3, 3), (4, 4), (9, 9), (6, 6), (6, 6), (5, 5), (4, 4), (0, 0), (7, 7), (4, 4), (0, 0), (1, 1), (3, 3), (1, 1), (3, 3), (4, 4), (7, 7), (2, 2), (7, 7), (1, 1), (3, 2), (1, 1), (1, 1), (7, 7), (4, 4), (2, 2), (3, 3), (5, 5), (1, 1), (2, 2), (4, 4), (4, 4), (6, 6), (3, 3), (5, 5), (5, 5), (6, 6), (0, 0), (4, 4), (1, 1), (9, 9), (5, 5), (7, 7), (8, 8), (9, 9), (2, 3), (7, 7), (4, 4), (6, 6), (4, 4), (3, 3), (0, 0), (7, 7), (0, 0), (2, 2), (9, 9), (1, 1), (7, 7), (3, 3), (2, 2), (9, 9), (7, 7), (7, 7), (6, 6), (2, 2), (7, 7), (8, 8), (4, 4), (7, 7), (3, 3), (6, 6), (1, 1), (3, 3), (6, 6), (9, 9), (3, 3), (1, 1), (4, 4), (1, 1), (7, 7), (6, 6), (9, 9)]


## Present the Results

Using logistic regression, we correctly recognize images in the test set ~92% of the time.